summarize.py

```diff
@@ -1,37 +1,57 @@
 from langchain.document_loaders import PyPDFLoader
 from langchain.text_splitter import CharacterTextSplitter
-from langchain.embeddings.openai import OpenAIEmbeddings
-from langchain.vectorstores import FAISS
-from langchain.chains import RetrievalQA
 from langchain import OpenAI
-
-
-def summarize(pdf_path):
+from langchain.chains.summarize import load_summarize_chain
+from time import time
+import argparse
+
+parser = argparse.ArgumentParser()
+parser.add_argument(
+    "-f",
+    "--file_path",
+    type=str,
+    help="Path to the PDF file to summarize",
+    required=True,
+)
+parser.add_argument(
+    "-c",
+    "--chain_type",
+    type=str,
+    help='Open ai summary chain_type: "stuff", "map_reduce" or "refine"',
+    default="map_reduce",
+)
+
+args = parser.parse_args()
+
+
+def summarize(pdf_path, chain_type):
     # Load the PDF
     print(f"Loading PDF from path: {pdf_path}")
-    loader = PyPDFLoader(file_path = pdf_path)
-
-    prompt = """Give me a 2 sentence summary of the following text: \n"""
+    loader = PyPDFLoader(file_path=pdf_path)
 
     # Split the PDF into chunks of text
-    text_splitter = CharacterTextSplitter(chunk_size=500, chunk_overlap=100, separator="/n")
+    text_splitter = CharacterTextSplitter(
+        chunk_size=750, chunk_overlap=100, separator="/n"
+    )
     document = loader.load()
-    docs = text_splitter.split_documents(documents = document)
-
-    # Create a vectorstore from the documents to use for retrieval
-    embeddings = OpenAIEmbeddings()
-    vectorstore =  FAISS.from_documents(docs, embeddings)
-    vectorstore.save_local("faiss_index_pdfs")
-    new_vectorstore = FAISS.load_local("faiss_index_pdfs", embeddings=embeddings)
-    
-    # Query OpenAI to summarize the docuemtns
-    qa = RetrievalQA.from_chain_type(llm=OpenAI(), chain_type="stuff", retriever = new_vectorstore.as_retriever())
+    docs = text_splitter.split_documents(documents=document)
 
-    openai_response = qa.run(prompt)
+    # Summarize throught openai
+    llm = OpenAI(temperature=0)
+    chain = load_summarize_chain(llm, chain_type=chain_type)
+    openai_response = chain.run(docs)
 
     return openai_response
 
 
 if __name__ == "__main__":
+    chain_type = "map_reduce"
     pdf_path = "/Users/jocelynmatsuo/devdir/lab/vectorsource-in-memory/2210.03629.pdf"
-    print(summarize(pdf_path=pdf_path))
\ No newline at end of file
+
+    start = (
+        time()
+    )  # timing the method, because certain chain_types take a super long time
```

Comment:
Which ones take a long time? Do you know why they take longer?

Reply:
the "refine" chain_type takes about 10 minutes - the calls are consecutive because each result gets used in the next call. "map_reduce" is about a second and a half. "Stuff" about the same, but almost useless, because best case, the first 4000 tokens are all it evaluates.